One problem with the current state of this PR that I can't really find a solution to is this:

The `isalpha` ufunc itself works correctly, both when called directly, and when called from the `np.char` module with an ndarray.

```python3
$ export PYTHONPATH="/Users/lysnikolaou/repos/python/numpy/build-install/usr/lib/python3.11/site-packages"
🐍 Launching Python with PYTHONPATH="/Users/lysnikolaou/repos/python/numpy/build-install/usr/lib/python3.11/site-packages"
$ /usr/bin/env python -P
Python 3.11.3 (tags/v3.11.3:f3909b8bc8, May 29 2023, 12:51:44) [Clang 14.0.3 (clang-1403.0.22.14.1)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy as np
>>> a = np.array([['abc', 'def'], ['world', ' he']])
>>> np.core.multi
np.core.multiarray np.core.multiply( 
>>> np.core.multiarray.isalpha(a)
array([[ True,  True],
       [ True, False]])
>>> np.char.isalpha(a)
array([[ True,  True],
       [ True, False]])
```

However, I'm getting a value error when passing a `chararray` to it.

```python3
$ export PYTHONPATH="/Users/lysnikolaou/repos/python/numpy/build-install/usr/lib/python3.11/site-packages"
🐍 Launching Python with PYTHONPATH="/Users/lysnikolaou/repos/python/numpy/build-install/usr/lib/python3.11/site-packages"
$ /usr/bin/env python -P
Python 3.11.3 (tags/v3.11.3:f3909b8bc8, May 29 2023, 12:51:44) [Clang 14.0.3 (clang-1403.0.22.14.1)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy as np
>>> a = np.char.array([['abc', 'def'], ['world', ' he']])
>>> np.core.multiarray.isalpha(a)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/lysnikolaou/repos/python/numpy/build-install/usr/lib/python3.11/site-packages/numpy/core/defchararray.py", line 2099, in __array_finalize__
    raise ValueError("Can only create a chararray from string data.")
ValueError: Can only create a chararray from string data.
```

It seems that the `bool` array returned from the ufuncs is somehow getting cast to a `chararray`, which is the input type. Is that something that should be happening? Am I missing something more obvious?

--- a/benchmarks/benchmarks/bench_core.py
+++ b/benchmarks/benchmarks/bench_core.py
@@ -247,3 +247,15 @@ class StatsMethods(Benchmark):
 
     def time_sum(self, dtype, size):
         self.data.sum()
+
+
+class NumPyChar(Benchmark):
+    def setup(self):
+        self.A = np.array([100*'x', 100*'y'])
+        self.B = np.array(1000 * ['aa'])
+
+    def time_isalpha_small_list_big_string(self):
+        np.char.isalpha(self.A)
+
+    def time_isalpha_big_list_small_string(self):
+        np.char.isalpha(self.B)

--- a/numpy/core/defchararray.py
+++ b/numpy/core/defchararray.py
@@ -2099,6 +2099,8 @@ class chararray(ndarray):
         return arr
 
     def __array_wrap__(self, arr, context=None):
+        if arr.dtype.char in "US":
+            return arr.view(type(self))
         return arr
 
     def __array_finalize__(self, obj):

--- a/numpy/core/defchararray.py
+++ b/numpy/core/defchararray.py
@@ -2093,6 +2093,14 @@ class chararray(ndarray):
         _globalvar = 0
         return self
 
+    def __array_prepare__(self, arr, context=None):
+        if arr.dtype.char in "US":
+            return arr.view(type(self))
+        return arr
+
+    def __array_wrap__(self, arr, context=None):
+        return arr
+
     def __array_finalize__(self, obj):
         # The b is a special case because it is used for reconstructing.
         if not _globalvar and self.dtype.char not in 'SUbc':

--- a/numpy/core/src/umath/string_ufuncs.cpp
+++ b/numpy/core/src/umath/string_ufuncs.cpp
@@ -129,6 +129,10 @@ string_isalpha(int len, const character *str)
         len = string_rstrip(len, str);
     }
 
+    if (len == 0) {
+        return (npy_bool) 0;
+    }
+
     for (int i = 0; i < len; i++) {
         npy_bool isalpha = (npy_bool) NumPyOS_ascii_isalpha(*str);
         if (!isalpha) {

